### PR TITLE
NumberLiteralSeparator: Stage 1 feature plugin. Closes gh-538

### PR DESCRIPTION
**Pull request by rwaldron**
Signed-off-by: Rick Waldron <waldron.rick@gmail.com>

| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | yes
| Deprecations?     | no
| Spec compliancy?  | yes
| Tests added/pass? | yes
| Fixed tickets     | #538
| License           | MIT

I modified `readInt` to ignore `_` characters found in decimal, hex, binary and octal literals. It will also throw SyntaxError exceptions when the `_` is misused.

### CODE REVIEW

**Comment by hzoo** [comment]
Amazing, that was super quick!

A few things to note:

- [ ] We need this to be under a flag like the other [plugins](https://github.com/babel/babylon#plugins). This just means creating a `hasPlugin('numericSeparators)` check so that this functionality isn't turned on by default.
  - You can check @diervo's PR as an example of this: https://github.com/babel/babylon/pull/260
- [ ] Update the readme to add that `numericSeparators` line to the plugins list
- [ ] Update the ast/spec.md (maybe no update needed)

**Comment by rwaldron** [comment]
@hzoo thanks for the quick review turnaround, I will get these issues addressed and ping you when ready :)

**Comment by hzoo** [comment]
Yeah, I should probably add that information to the contributing, let us know if there's anything that can be clarified or is frustrating

**Comment by rwaldron** [comment]
> - [x] We need this to be under a flag like the other plugins. This just means creating a hasPlugin('numericSeparators) check so that this functionality isn't turned on by default.
>   - You can check @diervo's PR as an example of this: #260
> - [x] Update the readme to add that numericSeparators line to the plugins list
> - [] ~~Update the ast/spec.md (maybe no update needed)~~

@hzoo changes pushed. I didn't make changes to ast/spec.md because I don't believe there are any changes to make there (I could be wrong, but need guidance)

File: src/tokenizer/index.js [hunk 1]
```diff
@@ -555,6 +572,23 @@ export default class Tokenizer extends LocationParser {
...
       let val;
+
+      if (this.hasPlugin("numericSeparator")) {
+        const prev = this.input.charCodeAt(this.state.pos - 1);
+        const next = this.input.charCodeAt(this.state.pos + 1);
+        if (code === 95) {
+          if (forbiddenNumericLiteralSeparatorSiblings.includes(prev) ||
+              forbiddenNumericLiteralSeparatorSiblings.includes(next) ||
+              Number.isNaN(next)) {
+            this.raise(this.state.pos, "Invalid NumericLiteralSeparator");
```
Comment (hzoo) [inline src/tokenizer/index.js]: I might actually create a different error message for multi `_`, if it's at the end or the beginning, or around invalid chars but we can do that later too.

File: src/tokenizer/index.js [hunk 2]
```diff
@@ -555,6 +572,23 @@ export default class Tokenizer extends LocationParser {
     for (let i = 0, e = len == null ? Infinity : len; i < e; ++i) {
       const code = this.input.charCodeAt(this.state.pos);
       let val;
+
+      if (this.hasPlugin("numericSeparator")) {
+        const prev = this.input.charCodeAt(this.state.pos - 1);
+        const next = this.input.charCodeAt(this.state.pos + 1);
+        if (code === 95) {
+          if (forbiddenNumericLiteralSeparatorSiblings.includes(prev) ||
```
Comment (DrewML) [inline src/tokenizer/index.js]: Thanks for the PR, @rwaldron! I believe we need to switch to `Array#indexOf` here. We don't load any polyfills in Babylon, and the minimum supported node version is 4.0.0.

Comment (hzoo) [inline src/tokenizer/index.js]: Dang forgot about that 🤔

Comment (rwaldron) [inline src/tokenizer/index.js]: Ah, ok, no problem. I saw all the other modern JS and just assumed I could use it...whoops! Will fix now.

**Comment by rwaldron** [comment]
@DrewML pushed change for includes -> indexOf

**Comment by rwaldron** [comment]
> Ideally the cases that have a weird error message should be fixed to have the same error message as the others

I believe most those stay as they are because they have a different violation in them before the illegal `_`, or are already illegal and I wanted to ensure that the new code didn't erroneously consider them a new kind of restriction.  

`._` is illegal today for 


I think this:

> 0o01_8 is Unexpected token, expected ; (1:5) (module)

...is an outlier that should be fixed, but I'm not sure how. The `_` is legal here, the `8` is not—it should have an error that says `Expected number in radix 8`. Seems like an unrelated issue for another ticket. 

And this: 

> ._1_1 is Unexpected token (1:0) (module)

... involves changing `readToken_dot` to raise an exception, which I could do here or in a follow up PR. A follow up might be nicer to keep the error message fixes together.

**Comment by hzoo** [comment]
Yeah I think this is good for this PR

**Comment by hzoo** [comment]
Let's do this 😄

**Comment by jridgewell** [comment]
Does `_123` pass?

**Comment by diervo** [comment]
It shouldn't the separator is only allowed in between numeric values.

**Comment by hzoo** [comment]
Well `_123` starts with _ so it should just be parsed as an Identifier right?

**Comment by rwaldron** [comment]
@jridgewell @hzoo @diervo correct, `_123` is not valid _DecimalDigits_, because `_` is a valid _IdentifierStart_ character. I'm sure it's fine, but I will open a PR that adds a test which asserts that this remains true when the `numericSeparator` plugin is active.

**Comment by michaelficarra** [comment]
@rwaldron I believe there's a bug in that this allows legacy octal integer literals to have separators. Even if this does restrict them, I don't see a test for that. Do you mind sending a PR to correct that?

**Comment by michaelficarra** [comment]
Ping @rwaldron.

**Pull request merged**